src/System.Diagnostics.Tracing/tests/BasicEventSourceTest/Harness/EventTestHarness.cs

```diff
@@ -0,0 +1,186 @@
+
+                    if (testNumber < tests.Count)
+                    {
+                        currentTest = tests[testNumber];
+                        Assert.Equal(currentTest.Name, data.PayloadValue(0, "name"));
+                        expectedTestNumber++;
+                        Console.WriteLine("Starting SubTest " + currentTest.Name);
+                    }
+                    else
+                    {
+                        Assert.NotNull(currentTest);
+                        Assert.Equal("", data.PayloadValue(0, "name"));
+                        Assert.Equal(tests.Count, testNumber);
+                        currentTest = null;
+                    }
+                }
+                else
+                {
+                    // If expectedTestNumber is 0 then this is before the first test
+                    // If expectedTestNumber is count then it is after the last test
+                    Assert.NotNull(currentTest);
+                    replies.Add(data);
+                }
+            };
+
+            // Run the tests. collecting and validating the results. 
+            using (TestHarnessEventSource testHarnessEventSource = new TestHarnessEventSource())
+            {
+                // Turn on the test EventSource.  
+                listener.EventSourceSynchronousEnable(source, options);
+                // And the harnesses's EventSource. 
+                listener.EventSourceSynchronousEnable(testHarnessEventSource);
+
+                // Generate events for all the tests, surrounded by events that tell us we are starting a test.  
+                int testNumber = 0;
+                foreach (var test in tests)
+                {
+                    testHarnessEventSource.StartTest(test.Name, testNumber);
+                    test.EventGenerator();
+                    testNumber++;
+                }
+                testHarnessEventSource.StartTest("", testNumber);        // Empty test marks the end of testing. 
+
+                // Disable the listeners.  
+                listener.EventSourceCommand(source.Name, EventCommand.Disable);
+                listener.EventSourceCommand(testHarnessEventSource.Name, EventCommand.Disable);
+
+                // Send something that should be ignored.  
+                testHarnessEventSource.IgnoreEvent();
+            }
+
+            Console.WriteLine("**** Disposing of the listener.");
+            listener.Dispose();         // Indicate we are done listening.  For the ETW file based cases, we do all the processing here
+
+            // expectedTetst number are the number of tests we successfully ran.  
+            Assert.Equal(expectedTestNumber, tests.Count);
+            Console.WriteLine("**** Ending Testing of " + tests.Count + " tests with Listener " + listener.GetType().Name);
+        }
+#region private
+        /// <summary>
+        /// This eventSource I use to emit events to separate tests from each other.  
```

Comment:
Nit: it'd be good to get rid of pronouns like "I" throughout the tests.  No one reading this code knows who "I" is.